vertx-sql-client-template/src/main/java/io/vertx/sqlclient/template/SqlTemplate.java

```diff
@@ -0,0 +1,145 @@
+   */
+  static SqlTemplate<Map<String, Object>, SqlResult<Void>> forUpdate(SqlClient client, String template) {
+    io.vertx.sqlclient.template.impl.SqlTemplate sqlTemplate = new io.vertx.sqlclient.template.impl.SqlTemplate((SqlClientInternal) client, template);
+    return new SqlTemplateImpl<>(client, sqlTemplate, query -> query.collecting(SqlTemplateImpl.NULL_COLLECTOR), Function.identity());
+  }
+
+  /**
+   * Set a parameters user defined mapping function.
+   *
+   * <p> At query execution, the {@code mapper} function is called to map the parameters object
+   * to a {@code Map<String, Object>} that configures the prepared query.
+   *
+   * @param mapper the mapping function
+   * @return a new template
+   */
+  <T> SqlTemplate<T, R> mapFrom(Function<T, Map<String, Object>> mapper);
+
+  /**
+   * Set a parameters user defined class mapping.
+   *
+   * <p> At query execution, the parameters object is is mapped to a {@code Map<String, Object>} that
+   * configures the prepared query.
+   *
+   * <p> This feature relies on {@link io.vertx.core.json.JsonObject#mapFrom} feature. This likely requires
+   * to use Jackson databind in the project.
+   *
+   * @param type the mapping type
+   * @return a new template
+   */
+  <T> SqlTemplate<T, R> mapFrom(Class<T> type);
+
+  /**
+   * Set a row user defined mapping function.
+   *
+   * <p> When the query execution completes, the {@code mapper} function is called to map the resulting
+   * rows to objects.
+   *
+   * @param mapper the mapping function
+   * @return a new template
+   */
+  <U> SqlTemplate<I, RowSet<U>> mapTo(Function<Row, U> mapper);
+
+  /**
+   * Set a row user defined mapping function.
+   *
+   * <p> When the query execution completes, resulting rows are mapped to {@code type} instances.
+   *
+   * <p> This feature relies on {@link io.vertx.core.json.JsonObject#mapFrom} feature. This likely requires
+   * to use Jackson databind in the project.
+   *
+   * @param type the mapping type
+   * @return a new template
+   */
+  <U> SqlTemplate<I, RowSet<U>> mapTo(Class<U> type);
+
+  /**
+   * Set a collector that will process the output and produce a custom result.
+   *
+   * @param collector the collector
+   * @return a new template
+   */
+  @GenIgnore
+  <U> SqlTemplate<I, SqlResult<U>> collecting(Collector<Row, ?, U> collector);
+
+  /**
+   * Execute the query with the {@code parameters}
+   *
+   * @param parameters the query parameters
+   * @param handler the result handler
+   */
+  void execute(I parameters, Handler<AsyncResult<R>> handler);
+
+  /**
+   * Like {@link #execute(Object, Handler)} but returns a {@code Future} of the asynchronous result
+   */
+  Future<R> execute(I params);
+
+  /**
+   * Execute a batch query with the {@code batch}.
+   *
+   * <p>Each item in the batch is mapped to a tuple.
+   *
+   * @param batch the batch
+   * @param handler the result handler
+   */
+  void executeBatch(List<I> batch, Handler<AsyncResult<R>> handler);
```

Comment:
because in practice the underlying query will actually use internally a `Promise<R>` that is created from the handler, so if you pass a promise to an `executeBatch` it will not be rewrapped into another promise.